runtime/codefrag.c

```diff
@@ -0,0 +1,126 @@
+                                enum digest_status digest_kind,
+                                unsigned char * opt_digest)
+{
+  int pos;
+  struct code_fragment * cf = caml_stat_alloc(sizeof(struct code_fragment));
+
+  cf->code_start = start;
+  cf->code_end = end;
+  switch (digest_kind) {
+  case DIGEST_LATER:
+    break;
+  case DIGEST_NOW:
+    caml_md5_block(cf->digest, cf->code_start, cf->code_end - cf->code_start);
+    digest_kind = DIGEST_PROVIDED;
+    break;
+  case DIGEST_PROVIDED:
+    memcpy(cf->digest, opt_digest, 16);
+    break;
+  case DIGEST_IGNORE:
+    break;
+  }
+  cf->digest_status = digest_kind;
+  caml_skiplist_insert(&caml_code_fragments_list,
+                       (uintnat) start, (uintnat) cf);
+  if (caml_code_fragments_table.capacity == 0) {
+    caml_ext_table_init(&caml_code_fragments_table, 8);
+  }
+  pos = caml_ext_table_add(&caml_code_fragments_table, cf);
+  cf->fragnum = pos;
+  return pos;
+}
+
+void caml_remove_code_fragment(struct code_fragment * cf)
+{
+  caml_skiplist_remove(&caml_code_fragments_list, (uintnat) cf->code_start);
+  /* Currently, caml_remove_code_fragment is used only to remove
+     the fragment most recently inserted, so optimize for this case. */
+  if (cf->fragnum == caml_code_fragments_table.size - 1) {
+    caml_code_fragments_table.size--;
+  } else {
+    /* Just mark the table entry invalid.  Later, we could maintain
+       a free list of empty entries in caml_code_fragments_table
+       so as to reuse them on the next caml_register_code_fragment.
+       Or replace the table by a skiplist. */
+    caml_code_fragments_table.contents[cf->fragnum] = NULL;
+  }
+  caml_stat_free(cf);
+}
+
+struct code_fragment * caml_find_code_fragment_by_pc(char *pc)
+{
+  struct code_fragment * cf;
+  uintnat key, data;
+
+  if (caml_skiplist_find_below(&caml_code_fragments_list,
+                               (uintnat) pc, &key, &data)) {
+    cf = (struct code_fragment *) data;
+    if (cf->code_start <= pc && pc < cf->code_end) return cf;
+  }
+  return NULL;
+}
+
+struct code_fragment * caml_find_code_fragment_by_num(int fragnum)
+{
+  if (0 <= fragnum && fragnum < caml_code_fragments_table.size) {
+    return caml_code_fragments_table.contents[fragnum];
+  } else {
+    return NULL;
+  }
+}
+
+unsigned char * caml_digest_of_code_fragment(struct code_fragment * cf)
+{
+  if (cf->digest_status == DIGEST_IGNORE)
+    return NULL;
+  if (cf->digest_status == DIGEST_LATER) {
+    caml_md5_block(cf->digest, cf->code_start, cf->code_end - cf->code_start);
+    cf->digest_status = DIGEST_PROVIDED;
+  }
+  return cf->digest;
+}
+
+struct code_fragment *
+   caml_find_code_fragment_by_digest(unsigned char digest[16])
+{
+  int i;
+
+  for (i = caml_code_fragments_table.size - 1; i >= 0; i--) {
+    struct code_fragment * cf = caml_code_fragments_table.contents[i];
+    unsigned char * d = caml_digest_of_code_fragment(cf);
+    if (d != NULL && memcmp(digest, d, 16) == 0) return cf;
+  }
```

Comment:
This loop is going to be executed each time a closure is deserialized. Hence, this might be performance sensitive. Why not also using a skiplist indexed on e.g., the first word of the MD5 sum? Since the skiplist implementation is already there, that should not be so difficult.

Reply:
Now that I think about it more, I understand that this might be more difficult than expected since several code fragments may share the same first word of digest.... We would need to patch the skiplist implementation to allow multiple bingdings for a key.

Reply:
You can link the code fragments that have a collision on the first word of the digest, like in a dynamic hash table.  But there's an additional wrinkle for code fragments whose digest is computed lazily.

Reply:
Why is this a problem? These code fragments would simply not appear in the skiplist.

Reply:
> These code fragments would simply not appear in the skiplist.

The DIGEST_IGNORE fragments would never be put in the skiplist (indexed by digests), yes.  The DIGEST_LATER fragments would need to be digested and inserted in the skiplist at the first call to `caml_find_code_fragment_by_digest`.

Reply:
The things you make me do, squared...  Have a look at commit 602aa0489 !

Reply:
Don't have a look at commit 602aa04 : removal is wrong.